pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -0,0 +1,429 @@
+			gClient, err := gcpclient.NewClient(cd.Spec.GCP.ProjectID, creds)
+			if err != nil {
+				reqLogger.Error(err, "Could not get gcp client with secret creds", "Secret Name", orgGcpSecretName, "Operator Namespace", operatorNamespace)
+				return reconcile.Result{}, err
+			}
+
+			// TODO(Raf) Check that operation is complete before continuing , make sure project Name does not exits , How to handle those errors
+			_, err = gClient.CreateProject(orgParentFolderID)
+			if err != nil {
+				reqLogger.Error(err, "Could create project", "Parent Folder ID", orgParentFolderID, "Requested Project Name", cd.Spec.Platform.GCP.ProjectID, "Requested Region Name", cd.Spec.GCP.Region)
+				return reconcile.Result{}, err
+			}
+
+			// TODO(Raf) Set quotas
+			// TODO(Raf) Enable APIs
+		} else {
+			// TODO(Raf) Should we requeue in this case or not ?
+			reqLogger.Error(err, "Could create project because cluster deployment does not have required fields", "Requested Project ID", cd.Spec.GCP.ProjectID, "Requested Region Name", cd.Spec.GCP.Region)
+			return reconcile.Result{}, err
+		}
+	}
+
+	gClient, err := gcpclient.NewClient(cd.Spec.GCP.ProjectID, creds)
+	if err != nil {
+		reqLogger.Error(err, "Could not get gcp client with secret creds", "Secret Name", orgGcpSecretName, "Operator Namespace", operatorNamespace)
+		return reconcile.Result{}, err
+	}
+
+	// See if service account exists if not create it
+	ServiceAccount, err := gClient.GetServiceAccount(osdServiceAccountName)
+	if err != nil {
+		// Create OSDManged Service account
+		Account, err := gClient.CreateServiceAccount(osdServiceAccountName, osdServiceAccountName)
+		if err != nil {
+			reqLogger.Error(err, "Could create service account", "Service Account Name", osdServiceAccountName)
+			return reconcile.Result{}, err
+		}
+		ServiceAccount = Account
+	}
+
+	// Configure policy
+	// Get policy from project
+	policy, err := gClient.GetIamPolicy()
+	if err != nil {
+		reqLogger.Error(err, "Could not get policy from project", "Project Name", cd.Spec.GCP.ProjectID)
+		return reconcile.Result{}, err
+	}
+
+	// Create requiredBindings with the new member
+	requiredBindings := GetOSDRequiredBindingMap(OSDRequiredRoles, ServiceAccount.Email)
+	// Get combined bindings
+	modified, newBindings := AddOrUpdateBinding(policy.Bindings, requiredBindings)
+	// If existing bindings have been modified update the policy
+	if modified {
+		// update policy
+		policy.Bindings = newBindings
+
+		setIamPolicyRequest := &cloudresourcemanager.SetIamPolicyRequest{
+			Policy: policy,
+		}
+
+		//TODO(Raf) Set Etag in policy to version policies so we get the latest always
+		_, err = gClient.SetIamPolicy(setIamPolicyRequest)
+		if err != nil {
+			reqLogger.Error(err, "Could not update policy on project", "Project Name", cd.Spec.GCP.ProjectID)
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Delete service account keys if any exist
+	err = gClient.DeleteServiceAccountKeys(ServiceAccount.Email)
```

Comment:
Why we doing this?

Reply:
We are doing this because if the reconcile fails after it creates a service account key and it goes back to loop it will not be able to access the key data. So in order to not store this in a temp secret, if a key already exists when we hit this part it will delete that one and create a new one and move to the next step.